codec-http/src/main/java/io/netty/handler/codec/http/HttpHeaderDateFormatter.java

```
@@ -0,0 +1,439 @@
+            char c1 = txt.charAt(tokenStart + 1);
+            if (!isDigit(c0) || !isDigit(c1)) {
+                return false;
+            }
+            year = getNumericalValue(c0) * 10 +
+                    getNumericalValue(c1);
+            return true;
+
+        } else if (len == 4) {
+            char c0 = txt.charAt(tokenStart);
+            char c1 = txt.charAt(tokenStart + 1);
+            char c2 = txt.charAt(tokenStart + 2);
+            char c3 = txt.charAt(tokenStart + 3);
+            if (!isDigit(c0) || !isDigit(c1) || !isDigit(c2) || !isDigit(c3)) {
+                return false;
+            }
+            year = getNumericalValue(c0) * 1000 +
+                    getNumericalValue(c1) * 100 +
+                    getNumericalValue(c2) * 10 +
+                    getNumericalValue(c3);
+            return true;
+        }
+
+        return false;
+    }
+
+    private void parseToken(CharSequence txt, int tokenStart, int tokenEnd) {
+        if (!timeFound) {
+            timeFound = tryParseTime(txt, tokenStart, tokenEnd);
+            if (timeFound) {
+                return;
+            }
+        }
+
+        if (!dayOfMonthFound) {
+            dayOfMonthFound = tryParseDayOfMonth(txt, tokenStart, tokenEnd);
+            if (dayOfMonthFound) {
+                return;
+            }
+        }
+
+        if (!monthFound) {
+            monthFound = tryParseMonth(txt, tokenStart, tokenEnd);
+            if (monthFound) {
+                return;
+            }
+        }
+
+        if (!yearFound) {
+            yearFound = tryParseYear(txt, tokenStart, tokenEnd);
+        }
+    }
+
+    private Date parse0(CharSequence txt, int start, int end) {
+        int tokenStart = -1;
+
+        for (int i = start; i < end; i++) {
+            char c = txt.charAt(i);
+
+            if (isDelim(c)) {
+                if (tokenStart != -1) {
+                    // terminate token
+                    parseToken(txt, tokenStart, i);
```

can we jump out early if we fail to parse a token, or we detect invalid input?